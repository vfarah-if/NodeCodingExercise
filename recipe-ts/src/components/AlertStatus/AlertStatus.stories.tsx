import { Story, Meta } from '@storybook/react/types-6-0'
import React from 'react'
import AlertStatus, { AlertStatusProps } from './AlertStatus'

export default {
  title: 'Component/AlertStatus',
  component: AlertStatus,
} as Meta

const AlertStatusTemplate: Story<AlertStatusProps> = args => (
  <>
    <AlertStatus {...args} />
  </>
)

export const InvisibleAlert = AlertStatusTemplate.bind({})
InvisibleAlert.args = {}

const error = new Error('Error Message')
export const AlertDefaultWithMessage = AlertStatusTemplate.bind({})
AlertDefaultWithMessage.args = {
  message: error.message,
}

export const AlertError = AlertStatusTemplate.bind({})
AlertError.args = {
  alertType: 'error',
  message: 'Error Message',
}

export const AlertSuccess = AlertStatusTemplate.bind({})
AlertSuccess.args = {
  alertType: 'success',
  message: 'Success messsage',
}

export const AlertInfo = AlertStatusTemplate.bind({})
AlertInfo.args = {
  alertType: 'info',
  message: 'Info messsage',
}

export const AlertWarning = AlertStatusTemplate.bind({})
AlertWarning.args = {
  alertType: 'warning',
  message: 'Warning messsage',
}
