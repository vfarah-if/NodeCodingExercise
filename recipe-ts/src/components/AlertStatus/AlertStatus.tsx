
import React from 'react';
import './style/index.css';

export interface AlertStatusProps {
    alertType?: 'error'|'success'|'info'|'warning', 
    message?: string | null
}

const AlertStatus: React.FC<AlertStatusProps> = ({alertType = 'error', message = ''}) => {
    // TODO: Try investigate how this can be fixed
    // const handleHideErrorAlert = (event:  React.MouseEvent<HTMLSpanElement, MouseEvent>) => {
    const handleHideErrorAlert = (event:  any) => {  
		const divElement = event.target.parentElement as HTMLDivElement;
		divElement.style.opacity = '0';
		setTimeout(() => (divElement.style.display = 'none'), 600);
	};

	return message ? (
		<div className={['alert', alertType].join(' ')}>
			<span className="closebtn" onClick={handleHideErrorAlert}>
				&times;
			</span>
			{message}
		</div>
	) : null;
}

export default AlertStatus;
