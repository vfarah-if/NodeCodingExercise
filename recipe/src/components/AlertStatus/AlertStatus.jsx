import React from 'react';
import PropTypes from 'prop-types';
import './style/index.css';

const AlertStatus = ({ alertType, message }) => {
	const handleHideErrorAlert = (event) => {
		const divElement = event.target.parentElement;
		divElement.style.opacity = '0';
		setTimeout(() => (divElement.style.display = 'none'), 600);
	};

	return message ? (
		<div className={['alert', alertType].join(' ')}>
			<span className="closebtn" onClick={handleHideErrorAlert}>
				&times;
			</span>
			{message}
		</div>
	) : null;
};

AlertStatus.propTypes = {
	alertType: PropTypes.oneOf(['error', 'success', 'info', 'warning']),
	message: PropTypes.string
};

AlertStatus.defaultProps = {
	alertType: 'error'
};

export default AlertStatus;