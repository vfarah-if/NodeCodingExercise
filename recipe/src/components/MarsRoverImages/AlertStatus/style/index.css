.alert {
	padding: 20px;
	background-color: #f44336;
	color: white;
	opacity: 1;
	transition: opacity 0.6s;
	margin-bottom: 15px;
}

.alert.error {
	background-color:  #f44336;
}

.alert.success {
	background-color: #4caf50;
}
.alert.info {
	background-color: #2196f3;
}
.alert.warning {
	background-color: #ff9800;
}

.closebtn {
	margin-left: 15px;
	color: white;
	font-weight: bold;
	float: right;
	font-size: 22px;
	line-height: 20px;
	cursor: pointer;
	transition: 0.3s;
}

.closebtn:hover {
	color: black;
}