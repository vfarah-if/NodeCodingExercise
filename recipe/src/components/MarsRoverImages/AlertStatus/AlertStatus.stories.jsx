
import React, {Fragment} from 'react';
import AlertStatus from './AlertStatus';

// eslint-disable-next-line import/no-anonymous-default-export
export default {
    title: 'Component/AlertStatus',
    component: AlertStatus
};

const AlertStatusTemplate = (args) => (
    <Fragment>
        <AlertStatus {...args} />
    </Fragment>
);

export const AlertErrorStory = AlertStatusTemplate.bind({});
AlertErrorStory.args = {
    alertType: 'error',
    message: 'Error Message'
}

export const AlertSuccessStory = AlertStatusTemplate.bind({});
AlertSuccessStory.args = {
    alertType: 'success',
    message: 'Success messsage'
}

export const AlertInfoStory = AlertStatusTemplate.bind({});
AlertInfoStory.args = {
    alertType: 'info',
    message: 'Info messsage'
}

export const AlertWarningStory = AlertStatusTemplate.bind({});
AlertWarningStory.args = {
    alertType: 'warning',
    message: 'Warning messsage'
}
