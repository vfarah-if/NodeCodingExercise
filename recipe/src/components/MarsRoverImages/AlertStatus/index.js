
import AlertStatus from './AlertStatus';
export default AlertStatus;    
