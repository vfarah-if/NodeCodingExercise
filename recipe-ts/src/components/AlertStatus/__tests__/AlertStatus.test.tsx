import React from 'react';
import { fireEvent, render, screen, waitForElementToBeRemoved } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';

import {
    AlertDefaultWithMessage,
    AlertError,
    AlertInfo,
    AlertSuccess,
    AlertWarning,
    InvisibleAlert
} from '../AlertStatus.stories';
import AlertStatus from '..';

describe('AlertStatus', () => {
	test('should create a default error alert with message', () => {
		const { container } = render(
			<AlertDefaultWithMessage
                alertType={AlertDefaultWithMessage.args?.alertType}
				message={AlertDefaultWithMessage.args?.message}
			/>
		);

		expect(container).toMatchSnapshot();
	});

	test('should create an alert error', () => {
		const { container } = render(
			<AlertError
                alertType={AlertError.args?.alertType}
				message={AlertError.args?.message}
			/>
		);

		expect(container).toMatchSnapshot();
	});

	test('should create an alert info', () => {
		const { container } = render(
			<AlertInfo
                alertType={AlertInfo.args?.alertType}
				message={AlertInfo.args?.message}
			/>
		);

		expect(container).toMatchSnapshot();
    });
    
    test('should create an alert success', () => {
		const { container } = render(
			<AlertSuccess
                alertType={AlertSuccess.args?.alertType}
				message={AlertSuccess.args?.message}
			/>
		);

		expect(container).toMatchSnapshot();
    });
    
    test('should create an alert warning', () => {
		const { container } = render(
			<AlertWarning
                alertType={AlertWarning.args?.alertType}
				message={AlertWarning.args?.message}
			/>
		);

		expect(container).toMatchSnapshot();
    });
    
    test('should create no alert when no message associated', () => {
		const { container } = render(
			<InvisibleAlert
                alertType={InvisibleAlert.args?.alertType}
				message={InvisibleAlert.args?.message}
			/>
		);

		expect(container).toMatchSnapshot();
    });
    
    test('should contain a close alert and should make alert invisible when clicked', async () => {
		const component = render(
			<AlertStatus alertType="error" message="Test Error" />
		);
		const linkElement = screen.queryByText('Ã—') as Element;
		expect(linkElement).toBeInTheDocument();

		// console.debug(screen.debug());
		expect(
			component.container.querySelector('div[style*="opacity: 0"]')
		).toBeNull();

		fireEvent.click(linkElement);

		expect(
			component.container.querySelector('div[style*="opacity: 0"]')
		).toBeDefined();
                
        // REMARK: Delays the timer however this is not correct
		// eslint-disable-next-line testing-library/await-async-utils
		await waitForElementToBeRemoved(() =>
			component.container.querySelector('div[style*="opacity: 0"]')
		).catch((err) => console.warn(err));        

		// console.debug(screen.debug());
		expect(
			component.container.querySelector(
				'div[style*="opacity: 0; display: none;"]'
			)
		).not.toBeNull();
	});
});
