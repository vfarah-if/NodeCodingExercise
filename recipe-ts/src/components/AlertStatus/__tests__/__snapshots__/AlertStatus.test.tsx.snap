// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AlertStatus should create a default error alert with message 1`] = `
<div>
  <div
    class="alert error"
  >
    <span
      class="closebtn"
    >
      ×
    </span>
    Error Message
  </div>
</div>
`;

exports[`AlertStatus should create an alert error 1`] = `
<div>
  <div
    class="alert error"
  >
    <span
      class="closebtn"
    >
      ×
    </span>
    Error Message
  </div>
</div>
`;

exports[`AlertStatus should create an alert info 1`] = `
<div>
  <div
    class="alert info"
  >
    <span
      class="closebtn"
    >
      ×
    </span>
    Info messsage
  </div>
</div>
`;

exports[`AlertStatus should create an alert success 1`] = `
<div>
  <div
    class="alert success"
  >
    <span
      class="closebtn"
    >
      ×
    </span>
    Success messsage
  </div>
</div>
`;

exports[`AlertStatus should create an alert warning 1`] = `
<div>
  <div
    class="alert warning"
  >
    <span
      class="closebtn"
    >
      ×
    </span>
    Warning messsage
  </div>
</div>
`;

exports[`AlertStatus should create no alert when no message associated 1`] = `<div />`;
